### PR TITLE
DC/OS BeakerX 0.16.1 with GPU Support

### PR DESCRIPTION
**Pull request by fabianbaier**

### CODE REVIEW

**Comment by fabianbaier** [comment]
```Diffing package beakerx revision 2 revision 3


Diffing File config.json:
6c6
<       "description": "DC/OS service configuration properties",
---
>       "description": "DC/OS BeakerX comes with a handful of pre-installed packages for data scientists. Since BeakerX 0.16.1 there is also GPU support included. Pre-Installed packages include: H5PY 2.7.1, Keras 2.1.5, Numpy 1.14.2, Pandas 0.22.0, PySpark 2.3.0, Scikit-Learn 0.19.1, TensorFlow & TensorBoard 1.7.0, XLRD 1.1.0",
9c9
<           "description": "Name of this service instance",
---
>           "description": "Name of this service instance.",
18a19,35
>         "gpu_support": {
>           "type": "object",
>           "description": "GPU support and useful packages for Data Scientists.\n\nGPU specific packages:\nCUDA 9.0.176, CUDNN 7.1.3.16, TensorFlow-GPU 1.7.0",
>           "properties": {
>             "enable": {
>               "description": "Enable GPU support.\nNote: This requires at least 15GB disk space available with recent NVIDIA drivers installed.",
>               "type": "boolean",
>               "default": false
>             },
>             "gpus": {
>               "description": "Number of GPUs to allocate to your DC/OS BeakerX  package.\nNote: GPU support has to be enabled.",
>               "type": "number",
>               "default": 0,
>               "minimum": 0
>             }
>           }
>         },
20c37
<           "description":  "Memory to allocate to each service instance.",
---
>           "description":  "Memory to allocate to each service instance in MB.",
33c50
<       "description": "beakerx storage configuration properties",
---
>       "description": "DC/OS BeakerX  storage configuration.",
36c53
<           "description": "Size of the persistent volume",
---
>           "description": "Size of the persistent volume in MB.",
55c72
<       "description": "beakerx networking configuration properties",
---
>       "description": "DC/OS BeakerX networking configuration properties",
57,61d73
<         "port_api": {
<           "description": "Port number to be used for api communication internally to the cluster.",
<           "type": "number",
<           "default": 8888
<         },
67c79
<               "description": "Enable or disable creating a VIP for external access through a public node running Marathon-LB.",
---
>               "description": "Enable or disable external access through a public node via Marathon-LB.",
72c84
<               "description": "For external access, dns to be used for Marathon-LB vHost: For example use your public slave elb dns.",
---
>               "description": "For external access, DNS to be used for Marathon-LB vHost: For example use your public slave ELB DNS.",
77c89
<               "description": "For external access, port number to be used for clear communication in the external Marathon-LB load balancer",
---
>               "description": "For external access, port number to be used for clear communication in the external Marathon-LB load balancer.",


Diffing File marathon.json.mustache:
3a4,6
> {{#service.gpu_support.enable}}
> "gpus": {{service.gpu_support.gpus}},
> {{/service.gpu_support.enable}}
12c15
< "VIP_0": "/{{service.name}}:{{networking.port_api}}"
---
> "VIP_0": "/{{service.name}}:8888"
30a34
> "volumes": [
32d35
<     "volumes": [
47d49
<     ],
48a51
> ],
49a53,56
> {{#service.gpu_support.enable}}
> "image": "{{resource.assets.container.docker.beakerx-docker-gpu}}",
> {{/service.gpu_support.enable}}
> {{^service.gpu_support.enable}}
50a58
> {{/service.gpu_support.enable}}
71a80,82
>   "uris": [
>     "{{resource.assets.uris.libmesos-bundle-tar-gz}}"
>   ],
91c102
< }
---
> }
\ No newline at end of file


Diffing File package.json:
2c2
<   "packagingVersion": "3.0",
---
>   "packagingVersion": "4.0",
5c5
<   "version": "0.12.2",
---
>   "version": "0.16.1-4",


Diffing File resource.json:
3,5c3,5
<     "icon-small": "https://media.licdn.com/mpr/mpr/shrink_200_200/AAEAAQAAAAAAAAOtAAAAJDg1NjRkMjdhLTU5YTktNDc3OC05Mzg5LWNhYjk0YjBiYjE2NQ.png",
<     "icon-medium": "https://media.licdn.com/mpr/mpr/shrink_200_200/AAEAAQAAAAAAAAOtAAAAJDg1NjRkMjdhLTU5YTktNDc3OC05Mzg5LWNhYjk0YjBiYjE2NQ.png",
<     "icon-large": "https://media.licdn.com/mpr/mpr/shrink_200_200/AAEAAQAAAAAAAAOtAAAAJDg1NjRkMjdhLTU5YTktNDc3OC05Mzg5LWNhYjk0YjBiYjE2NQ.png",
---
>     "icon-small": "https://downloads.mesosphere.com/beakerx/assets/beakerx-logo.png",
>     "icon-medium": "https://downloads.mesosphere.com/beakerx/assets/beakerx-logo.png",
>     "icon-large": "https://downloads.mesosphere.com/beakerx/assets/beakerx-logo.png",
7c7
<       "https://user-images.githubusercontent.com/963093/28300136-585f9f7c-6b4b-11e7-8827-b5807d3fc9a8.png"
---
>       "https://downloads.mesosphere.com/beakerx/assets/beakerx-screenshot.png"
10a11,13
>     "uris": {
>       "libmesos-bundle-tar-gz": "https://downloads.mesosphere.com/libmesos-bundle/libmesos-bundle-1.11.0.tar.gz"
>     },
13c16,17
<         "beakerx-docker": "mesosphere/dcos-beakerx:0.12.2"
---
>         "beakerx-docker": "mesosphere/dcos-beakerx:0.16.1-base",
>         "beakerx-docker-gpu": "mesosphere/dcos-beakerx:0.16.1-gpu"
```

File: repo/packages/B/beakerx/3/config.json [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        },
+        "cpus": {
+          "description": "CPU shares to allocate to each service instance.",
+          "type": "number",
+          "default": 1,
+          "minimum": 1
+        },
+        "gpu_support": {
+          "type": "object",
+          "description": "GPU support and useful packages for Data Scientists.\n\nAdditional packages installed:\nCUDA 9.0.176, CUDNN 7.1.3.16, Tensorflow-GPU 1.7.0",
```
Comment (joerg84) [inline repo/packages/B/beakerx/3/config.json]: GPU support doesn't add more packages or?

File: repo/packages/B/beakerx/3/config.json [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+              "type": "boolean",
+              "default": false
+            },
+            "mount_agent_libraries": {
+              "description": "Enable native support of Spark 2.2.1 in your DC/OS cluster.\nNote: Jupyter will run under root and the container has access to /opt/mesosphere to utilize your agents libraries.",
+              "type": "boolean",
+              "default": false
+            },
+            "gpus": {
+              "description": "GPU shares to allocate to your Beakerx package.\nNote: GPU support has to be enabled.",
```
Comment (joerg84) [inline repo/packages/B/beakerx/3/config.json]: Is this full GPU?

File: repo/packages/B/beakerx/3/config.json [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+              "type": "number",
+              "default": 0,
+              "minimum": 0
+            }
+          }
+        },
+        "mem": {
+          "description":  "Memory to allocate to each service instance.",
```
Comment (joerg84) [inline repo/packages/B/beakerx/3/config.json]: in MB

File: repo/packages/B/beakerx/3/config.json [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        }
+      },
+      "required": [
+        "cpus",
+        "mem"
+      ]
+    },
+    "storage": {
+      "type": "object",
+      "description": "beakerx storage configuration properties",
```
Comment (joerg84) [inline repo/packages/B/beakerx/3/config.json]: Capitalization, in MB?

File: repo/packages/B/beakerx/3/config.json [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+              "type": "boolean",
+              "default": false
+            }
+          }
+        }
+      }
+    },
+    "networking": {
+      "type": "object",
+      "description": "beakerx networking configuration properties",
```
Comment (joerg84) [inline repo/packages/B/beakerx/3/config.json]: Capitalization

File: repo/packages/B/beakerx/3/config.json [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+          "description": "Port number to be used for api communication internally to the cluster.",
+          "type": "number",
+          "default": 8888
+        },
+        "external_access": {
+          "type": "object",
+          "description": "Enable access from outside the cluster through Marathon-LB.\nNOTE: this connection is unencrypted.",
+          "properties": {
+            "enable": {
+              "description": "Enable or disable creating a VIP for external access through a public node running Marathon-LB.",
```
Comment (joerg84) [inline repo/packages/B/beakerx/3/config.json]: IMO this is not a VIP.... but the VHOST entry for HA_PROXY

File: repo/packages/B/beakerx/3/resource.json [hunk 7]
```diff
@@ -0,0 +1,18 @@
+{
+  "images": {
+    "icon-small": "https://secure.gravatar.com/avatar/35c6fa7c0a103dc1cfa31b7ff56215fb.jpg",
```
Comment (joerg84) [inline repo/packages/B/beakerx/3/resource.json]: Haven't we moved all images to our s3 downloads?

File: repo/packages/B/beakerx/3/resource.json [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    "icon-medium": "https://secure.gravatar.com/avatar/35c6fa7c0a103dc1cfa31b7ff56215fb.jpg",
+    "icon-large": "https://secure.gravatar.com/avatar/35c6fa7c0a103dc1cfa31b7ff56215fb.jpg",
+    "screenshots": [
+      "https://user-images.githubusercontent.com/963093/28300136-585f9f7c-6b4b-11e7-8827-b5807d3fc9a8.png"
```
Comment (joerg84) [inline repo/packages/B/beakerx/3/resource.json]: same as above.

**Comment by fabianbaier** [comment]
sgtm, addressed review.

File: repo/packages/B/beakerx/3/config.json [hunk 9]
```diff
@@ -0,0 +1,108 @@
+{
+  "type": "object",
+  "properties": {
+    "service": {
+      "type": "object",
+      "description": "DC/OS BeakerX comes with a handful of pre-installed packages for data scientists. Since BeakerX 0.16.1 there is also GPU support included. Pre-Installed packages include: H5PY 2.7.1, Keras 2.1.5, Numpy 1.14.2, Pandas 0.22.0, PySpark 2.3.0, Scikit-Learn 0.19.1, Tensorflow & Tensorboard 1.7.0, XLRD 1.1.0",
```
Comment (joerg84) [inline repo/packages/B/beakerx/3/config.json]: TensorBoard

File: repo/packages/B/beakerx/3/config.json [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+      "properties": {
+        "name": {
+          "description": "Name of this service instance",
```
Comment (joerg84) [inline repo/packages/B/beakerx/3/config.json]: instance*.*

File: repo/packages/B/beakerx/3/config.json [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+          "type": "object",
+          "description": "GPU support and useful packages for Data Scientists.\n\nGPU specific packages:\nCUDA 9.0.176, CUDNN 7.1.3.16, Tensorflow-GPU 1.7.0",
+          "properties": {
+            "enable": {
+              "description": "Enable GPU support.\nNote: This requires at least 15GB disk space available with recent NVIDIA drivers installed.",
+              "type": "boolean",
+              "default": false
+            },
+            "mount_agent_libraries": {
+              "description": "Enable native support of Spark 2.2.1 in your DC/OS cluster.\nNote: Jupyter will run under root and the container has access to /opt/mesosphere to utilize your agents libraries.",
```
Comment (joerg84) [inline repo/packages/B/beakerx/3/config.json]: Is it always that Spark version?

File: repo/packages/B/beakerx/3/marathon.json.mustache [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    "containerPath": "persistent_data"
+    },
+    {
+    "containerPath": "/home/beakerx/persistent_data",
+    "hostPath": "persistent_data",
+    "mode": "RW"
+    }
+{{#service.gpu_support.mount_agent_libraries}}
+      ,{
+        "containerPath": "/opt/mesosphere",
```
Comment (joerg84) [inline repo/packages/B/beakerx/3/marathon.json.mustache]: Do we mount the entire /opt/mesosphere somewhere else?

File: repo/packages/B/beakerx/3/package.json [hunk 13]
```diff
@@ -0,0 +1,20 @@
+{
+  "packagingVersion": "4.0",
+  "minDcosReleaseVersion": "1.9",
+  "name": "beakerx",
+  "version": "0.16.1",
```
Comment (joerg84) [inline repo/packages/B/beakerx/3/package.json]: I would make the version in the format `baseVersion`-`packageVersion`, i.e., 0.16.1-1

File: repo/packages/B/beakerx/3/config.json (hunk 11 above)
Comment (joerg84) [reply repo/packages/B/beakerx/3/config.json]: Please remove as discussed

File: repo/packages/B/beakerx/3/marathon.json.mustache (hunk 12 above)
Comment (fabianbaier) [reply repo/packages/B/beakerx/3/marathon.json.mustache]: We used to e.g. with Jenkins. But as discussed I remove this option in favor of providing the URI

**Review by joerg84 — APPROVED** [review]

**Pull request merged**